File: app/src/main/java/com/softaai/designpatterns/java/creational/AbstractFactory/AbstractFactoryPatternDemo.java
package com.softaai.designpatterns.java.creational.AbstractFactory;

import com.softaai.designpatterns.java.creational.AbstractFactory.color.Color;
import com.softaai.designpatterns.java.creational.AbstractFactory.color.ColorFactory;
import com.softaai.designpatterns.java.creational.AbstractFactory.shape.Shape;

/**
 * Created by Amol Pawar on 31-03-2019.
 * softAai Apps
 */
public class AbstractFactoryPatternDemo {

    public static void main(String[] args){

        AbstractFactory shapeFactory = FactoryProducer.getFactory("Shape");

        Shape rectangle = shapeFactory.getShape("Rectangle");
        rectangle.draw();
        Shape square = shapeFactory.getShape("square");
        square.draw();
        Shape circle = shapeFactory.getShape("Circle");
        circle.draw();


        AbstractFactory colorFactory = FactoryProducer.getFactory("Color");

        Color red = colorFactory.getColor("Red");
        red.fill();
        Color blue = colorFactory.getColor("Blue");
        blue.fill();
        Color green = colorFactory.getColor("green");
        green.fill();
    }
}

File: app/src/main/java/com/softaai/designpatterns/kotlin/creational/AbstractFactory/AbstractFactoryPatternDemo.kt
package com.softaai.designpatterns.kotlin.creational.AbstractFactory

import com.softaai.designpatterns.kotlin.creational.AbstractFactory.color.Color
import com.softaai.designpatterns.kotlin.creational.AbstractFactory.shape.Shape


/**
 * Created by Amol Pawar on 01-04-2019.
 * softAai Apps
 */
object AbstractFactoryPatternDemo {
    @JvmStatic    // we can use this with either object or companion object and for testing purpose we use only
    fun main(args:Array<String>){

       val shapeFactory:AbstractFactory?
       shapeFactory = FactoryProducer.getFactory("Shape")

       val rectangle: Shape?
       rectangle = shapeFactory?.getShape("Rectangle")
       rectangle?.draw()

       val square:Shape?
       square = shapeFactory?.getShape("Square")
       square?.draw()

       val circle:Shape?
       circle = shapeFactory?.getShape("Circle")
       circle?.draw()



       val colorFactory:AbstractFactory?
       colorFactory = FactoryProducer.getFactory("Color")

       val red: Color?
       red = colorFactory?.getColor("RED")
       red?.fill()

       val blue: Color?
       blue = colorFactory?.getColor("BLUE")
       blue?.fill()

       val green: Color?
       green = colorFactory?.getColor("GREEN")
       green?.fill()

    }
}